runtime/moonbeam/src/lib.rs

```diff
@@ -1367,6 +1384,44 @@ impl pallet_parameters::Config for Runtime {
 	type WeightInfo = moonbeam_weights::pallet_parameters::WeightInfo<Runtime>;
 }
 
+impl pallet_multiblock_migrations::Config for Runtime {
+	type RuntimeEvent = RuntimeEvent;
+	// TODO fully replace pallet_migrations with multiblock migrations.
+	#[cfg(not(feature = "runtime-benchmarks"))]
+	type Migrations = pallet_identity::migration::v2::LazyMigrationV1ToV2<Runtime>;
+	// Benchmarks need mocked migrations to guarantee that they succeed.
+	#[cfg(feature = "runtime-benchmarks")]
+	type Migrations = pallet_multiblock_migrations::mock_helpers::MockedMigrations;
+	type CursorMaxLen = ConstU32<65_536>;
+	type IdentifierMaxLen = ConstU32<256>;
+	type MigrationStatusHandler = ();
+	type FailedMigrationHandler = ActivateMaintenanceModeOnFailedMigration;
+	type MaxServiceWeight = MaxServiceWeight;
+	type WeightInfo = weights::pallet_multiblock_migrations::WeightInfo<Runtime>;
+}
+
+/// A `FailedMigrationHandler` that enables maintenance mode.
+pub struct ActivateMaintenanceModeOnFailedMigration;
+impl frame_support::migrations::FailedMigrationHandler
+	for ActivateMaintenanceModeOnFailedMigration
+{
+	fn failed(migration: Option<u32>) -> frame_support::migrations::FailedMigrationHandling {
+		// Log information about the failed migration
+		log::error!(
+			target: "runtime::migrations",
+			"Migration {:?} failed - activating maintenance mode and continuing",
+			migration
+		);
+
+		// Attempt to enable maintenance mode.
+		let _ = MaintenanceMode::enter_maintenance_mode(RuntimeOrigin::root());
```

Comment:
It will not work because `MaintenanceOrigin` doesn't allow root.
Also, error cases are not handle here.
it might be better to add a public function in maintenance pallet (moonkit) and call it instead of the dispatchable one.

Reply:
Addressed in:
- moonkit: https://github.com/Moonsong-Labs/moonkit/commit/e941b9009862105a0e93e8163d027c57013a3082
- moonbeam: https://github.com/moonbeam-foundation/moonbeam/pull/3224/commits/7f73312e84b5795f92c1b188e0fea43bcf83b9c5